[E2E Test] Utilities for State Sync Test

Olshansk opened the pull request:
## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 03 Aug 23 20:11 UTC
This pull request introduces several changes across multiple files. 

Here is a summary of the changes:

1. `go.mod`:
   - The `golang.org/x/text` dependency is now listed explicitly instead of being an indirect dependency.

2. `scenario_test.go`:
   - Added a new feature called "State Sync Namespace" that includes various commands and waits for specific amounts of time.
   - Code improvements and TODO comments have been added.

3. `FAQ.md`:
   - Updated an issue related to starting LocalNet with SELinux on an operating system. Replaced the command `make compose_and_watch` with `make lightweight_localnet` to avoid permission denied errors.

4. `.gitignore`:
   - Removed the entry "main" from the list of ignored files.
   - Removed the entries "rpc/server.gen.go" and "rpc/client.gen.go" from the list of ignored files.
   - Added the entry "**/gomock_reflect_*/" to ignore mock temporary files.

5. `e2e/README.md`:
   - Added a new section on `Keywords`.
   - Modified scenario descriptions and code examples to replace instances of "Validator" with "Node".
   - Included a flowchart depicting the E2E scenarios with updated terminology.

6. Consensus module files:
   - Added a new logging statement in the `HandleDebugMessage` function.
   - Simplified the handling of the `DEBUG_CONSENSUS_RESET_TO_GENESIS` action.

7. `.tiltignore`:
   - Removed the entry "main" from the list of ignored files.
   - Removed the entries "rpc/server.gen.go" and "rpc/client.gen.go" from the list of ignored files.

8. `CHANGELOG.md`:
   - Updated build commands and added a new section on `Keywords`.

9. `persistence/docs/CHANGELOG.md`:
   - Several changes related to deprecation, addition, and fixing of functions and issues.

10. `validator.feature`:
    - Renamed file from "valdator.feature" to "validator.feature".
    - Updated scenario titles and step descriptions to use more descriptive terminology.
    - Replaced references to "validator" with "node".

11. `tilt_helpers.go`:
    - Added a new file containing functions related to syncing network configuration and checking package installation.

12. `debug.go`:
    - Added new debug commands and subcommands.
    - Updated existing functions and added new functions for debug actions.
    
13. `account.feature`:
    - Added a new file containing scenarios for testing node account functionalities.

14. `README.md` files:
    - Updated sections, titles, dependencies, and instructions in various README.md files.

15. `build/config/README.md`:
    - Updated usage instructions, changing the command `make compose_and_watch` to `make lightweight_localnet`.

16. `iteration_3_end_to_end_tx.md`:
    - Updated commands to start LocalNet and consensus debugger.

17. Deleted files:
   - `validator.go`
   - `watch_build.sh`

18. New files added:
   - `debug.feature`
   - `tilt_helpers.go`
   - `account.feature`

Please let me know if you need more information about any specific change.
<!-- reviewpad:summarize:end -->

## Issue

Fixes par of #579

## Type of change

Please mark the relevant option(s):

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other <!-- add details here if it a different type of change -->

## List of changes

- `s/compose_and_watch/lightweight_localnet` and all related helpers
- `s/validator/node` in e2e tests for clarity
- Add fire-and-forget `Debug` CLI w/ several useful initial subcommands
- Add `keywords` to the `e2e` document
- Add an `e2e debug` test to trigger views and track the blockchain increasing
- Avoid rebuilding the actors if the CLI changes
- Small miscellaneous improvements & code cleanup

## Testing

- [x] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [ ] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [ ] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [x] I have updated the corresponding README(s); local and/or global
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

Olshansk commented:
@dylanlott @red-0ne @adshmh Please note that this test is still a bit flaky for now (will solidify it next week), but I think this change is close enough to start getting reviews on.

dylanlott commented on e2e/tests/debug.feature:
```diff
@@ -0,0 +1,16 @@
+    # Since the configuration for consensus is optimistically responsive, we need to be in manual
+    # Pacemaker mode and call TriggerView to further the blockchain.
+    # 1 second was chosen arbitrarily for the time for block propagation.
+    Scenario: 4 Validator blockchain from genesis reaches block 2 when TriggerView is executed twice
```
:100:

dylanlott commented on build/localnet/Tiltfile:
```diff
@@ -2,7 +2,8 @@
 load("ext://restart_process", "docker_build_with_restart")
-load('ext://tests/golang', 'test_go')
+load("ext://tests/golang", "test_go")
+load("ext://syncback", "syncback")
```
This broke my localnet because I didn't have an update to date (>3.0.0 or greater) version of rsync. The dependency should be listed in the prerequisites section of the development setup guide.
More info at [tilt-extensions/syncback - Usage](https://github.com/tilt-dev/tilt-extensions/tree/master/syncback#usage)

Olshansk replied:
[`489978f` (#874)](https://github.com/pokt-network/pocket/pull/874/commits/489978f040cffab5b146014ecaf12222f5828e70)

dylanlott commented on e2e/tests/steps_init_test.go:
```diff
@@ -71,24 +72,136 @@ func TestFeatures(t *testing.T) {
+		actor,
+	}
+
+	// Depending ont he type of `actor` we're querying, we'll have a different expected responses
```
```suggestion
	// Depending on the type of `actor` we're querying, we'll have a different set of expected responses
```

dylanlott commented on app/client/cli/debug.go:
```diff
@@ -34,43 +39,123 @@ var items = []string{
+			Args:              cobra.ExactArgs(0),
 			PersistentPreRunE: helpers.P2PDependenciesPreRunE,
-			Run: func(cmd *cobra.Command, _ []string) {
-				handleSelect(cmd, cmd.Use)
```
Glad to see this finally refactored into appropriate Viper subcommands 🐍

0xBigBoss reviewed (commented):
I didn't review everything in detail but this approach for handling e2e looks great. I was taking a bit of a different approach instead of waiting for a specified time after triggering the next view, I added more state to the test suite and track pending transaction hashes waiting for them to be all committed. Using the tilt executable too is a nice touch for introducing more participants.

potentially instead of a simple wait, perhaps we can add a step that gets the current height, triggers next view, and then polls for the height to change.

All in all in this is a great direction for the stability of the e2e tests. 

See https://github.com/pokt-network/pocket/pull/948/files#diff-01dec4121ae8acb7a1f4bb72a6c2104827d2c2d2197eb5f45fa5c032ffba32cdR93

Olshansk commented:
@red-0ne Could you please prioritize taking a look at this?

Olshansk commented:
@0xBigBoss I've created #959 and added TODO comments so we can either poll on heights on healthchecks in a followup commit. Also, to avoid PRs that are too large, would appreciate if you could split out things that are possible.

For example, this used to be part of the state sync PR, but I split it out to make it easier for the reviewer instead of having one multi-thousand line PR.

red-0ne commented on e2e/tests/tilt_helpers.go:
```diff
@@ -0,0 +1,34 @@
+		e2eLogger.Debug().Msgf("syncLocalNetConfigFromHostToLocalFS: 'tilt' is not installed, skipping...")
+		return
+	}
+	sedCmd := exec.Command("tilt", "trigger", "syncback_localnet_config")
```
Does this variable name mean anything specific?

red-0ne commented on e2e/tests/node.go:
```diff
@@ -0,0 +1,76 @@
+	n.result = r
+	// IMPROVE: make targetPodName configurable
+	n.targetPodName = targetDevClientPod
+	if err != nil {
```
Shouldn't this be right after `out, err :=cmd.Output()`?

red-0ne reviewed (approved):
Have minor comments.
It's a good base to test state sync against. LGTM :+1: 

Actually, I'm having trouble running making the e2e test succeed. Will comeback after resolving it.

0xBigBoss commented on build/scripts/watch.sh:
```diff
@@ -19,7 +19,8 @@ else
-  --decoration="none" \
-  -s -- sh -c "$command";
+    --start-service \
+    -R '^app/client' \
```
ty! lol was driving me crazy when working on my debug client.

0xBigBoss reviewed (approved):
all in all a great base addition to the e2e tests. i think it's good to merge and work on removing the timers in separate PRs.

Olshansk commented on e2e/tests/state_sync.feature:
```diff
@@ -0,0 +1,23 @@
+        # full_nodes is the key used in `localnet_config.yaml`
+        When the developer runs the command "ScaleActor full_nodes 2"
+        # IMPROVE: Figure out if there's something better to do then waiting for a node to spin up
+        And the developer waits for "40000" milliseconds
```
https://github.com/pokt-network/pocket/issues/959#issuecomment-1664476974
